File: frontend/src/api/files.ts
import { get, post } from "./client";

export interface TreeNode {
  name: string;
  path: string;
  is_dir: boolean;
  children?: TreeNode[] | null;
}

export function getTree() {
  return get<TreeNode>("/api/files/tree");
}

export async function getFileTreeChildren(path: string): Promise<TreeNode[]> {
  const res = await get<{ children: TreeNode[] }>("/api/files/children", { path });
  return res.children ?? [];
}

export async function readFileText(path: string): Promise<string> {
  const url = new URL("/api/files/read", window.location.origin);
  url.searchParams.set("path", path);
  url.searchParams.set("as_text", "true");
  const res = await fetch(url.pathname + url.search);
  if (!res.ok) throw new Error(await res.text());
  return res.text();
}

export async function writeFile(path: string, content: string): Promise<void> {
  const res = await fetch("/api/files/write", {
    method: "PUT",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ path, content }),
  });
  if (!res.ok) throw new Error(await res.text());
}

export async function createFile(path: string, content = ""): Promise<void> {
  const res = await fetch("/api/files/file", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ path, content }),
  });
  if (!res.ok) throw new Error(await res.text());
}

export async function createDir(path: string): Promise<void> {
  const res = await fetch("/api/files/dir", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ path }),
  });
  if (!res.ok) throw new Error(await res.text());
}

export async function deletePath(path: string): Promise<void> {
  const res = await fetch(`/api/files?path=${encodeURIComponent(path)}`, {
    method: "DELETE",
  });
  if (!res.ok) throw new Error(await res.text());
}

export async function movePath(fromPath: string, toPath: string): Promise<void> {
  const res = await fetch("/api/files/move", {
    method: "PATCH",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ from_path: fromPath, to_path: toPath }),
  });
  if (!res.ok) throw new Error(await res.text());
}

export function backupDatas() {
  return post<{ files_backed_up: number }>("/api/backup/datas");
}

File: backend/api/files.py
"""File and backup API."""
from fastapi import APIRouter, HTTPException
from fastapi.responses import Response, PlainTextResponse

from core.exceptions import PathOutsideWorkspaceError, NotFoundError
from pydantic import BaseModel
from services import file_service

router = APIRouter(prefix="/api/files", tags=["files"])


class WriteFileBody(BaseModel):
    path: str
    content: str


class CreateFileBody(BaseModel):
    path: str
    content: str = ""


class CreateDirBody(BaseModel):
    path: str


class MoveBody(BaseModel):
    from_path: str
    to_path: str


@router.get("/tree")
def get_tree():
    tree = file_service.get_tree()
    return tree


@router.get("/children")
def list_children(path: str = ""):
    """List direct children of path (for expanding tree)."""
    try:
        nodes = file_service.list_dir(path)
        return {"children": [n.model_dump() for n in nodes]}
    except PathOutsideWorkspaceError as e:
        raise HTTPException(403, str(e.message))


@router.get("/read")
def read_file(path: str = "", as_text: bool = False):
    if not path:
        raise HTTPException(400, "path required")
    try:
        raw = file_service.read_file(path)
    except PathOutsideWorkspaceError as e:
        raise HTTPException(403, str(e.message))
    except NotFoundError as e:
        raise HTTPException(404, str(e.message))
    if as_text:
        try:
            text = raw.decode("utf-8")
        except UnicodeDecodeError:
            raise HTTPException(400, "File is not valid UTF-8 text")
        return PlainTextResponse(content=text)
    return Response(content=raw, media_type="application/octet-stream")


@router.put("/write")
def write_file_endpoint(body: WriteFileBody):
    if not body.path or not body.path.strip():
        raise HTTPException(400, "path required")
    try:
        file_service.write_file(body.path.strip(), body.content.encode("utf-8"))
    except PathOutsideWorkspaceError as e:
        raise HTTPException(403, str(e.message))
    except NotFoundError as e:
        raise HTTPException(404, str(e.message))
    return {"ok": True}


@router.post("/file")
def create_file_endpoint(body: CreateFileBody):
    if not body.path or not body.path.strip():
        raise HTTPException(400, "path required")
    try:
        file_service.create_file(body.path.strip(), body.content.encode("utf-8"))
    except PathOutsideWorkspaceError as e:
        raise HTTPException(403, str(e.message))
    except NotFoundError as e:
        raise HTTPException(404, str(e.message))
    return {"ok": True}


@router.post("/dir")
def create_dir_endpoint(body: CreateDirBody):
    if not body.path or not body.path.strip():
        raise HTTPException(400, "path required")
    try:
        file_service.create_dir(body.path.strip())
    except PathOutsideWorkspaceError as e:
        raise HTTPException(403, str(e.message))
    return {"ok": True}


@router.delete("")
def delete_path_endpoint(path: str = ""):
    if not path or not path.strip():
        raise HTTPException(400, "path required")
    try:
        file_service.delete_path(path.strip())
    except PathOutsideWorkspaceError as e:
        raise HTTPException(403, str(e.message))
    except NotFoundError as e:
        raise HTTPException(404, str(e.message))
    return {"ok": True}


@router.patch("/move")
def move_path_endpoint(body: MoveBody):
    if not body.from_path or not body.to_path:
        raise HTTPException(400, "from_path and to_path required")
    try:
        file_service.rename_path(body.from_path.strip(), body.to_path.strip())
    except PathOutsideWorkspaceError as e:
        raise HTTPException(403, str(e.message))
    except NotFoundError as e:
        raise HTTPException(404, str(e.message))
    return {"ok": True}



